### PR TITLE
Add less and greater filters to DB views

### LINKED ISSUE #187: Add less and greater filters to DB views
### Description
**Pull Request:** #177 
This need arose during the development of the **Cohort Analysis** project.

Add **less than**, **less than equal**, **greater than**, and **greater than equals** filters to the DB views.
This would make it possible to filter for items that are younger than 30 years of age.

The parameter to use the filters are:
* less than = `filter_lt`
* greater than = `filter_gt`
* less than equal = `filter_lte`
* greater than equal = `filter_gte`

### Examples
1. **filter_lte:** Filter for all white females that are 25 years old and younger
http://localhost:8080/api/tdp/db/publicdb/tissue?filter_gender=female&filter_race=white&filter_lte_age=25
2. **filter_lt:** Filter for all white females that are younger than 25 years
http://localhost:8080/api/tdp/db/publicdb/tissue?filter_gender=female&filter_race=white&filter_lt_age=25
3. **filter_gte:** Filter for all white females that are 88 years old and older
http://localhost:8080/api/tdp/db/publicdb/tissue?filter_gender=female&filter_race=white&filter_gte_age=88
4. **filter_lt:** Filter for all white females that are older than 88 years
http://localhost:8080/api/tdp/db/publicdb/tissue?filter_gender=female&filter_race=white&filter_gt_age=88
5. **Combination** of 2 filters on the **same column**: Filter for all white females that are 15 years and older as well as 21 years and younger
http://localhost:8080/api/tdp/db/publicdb/tissue?filter_gender=female&filter_race=white&filter_gte_age=15&filter_lte_age=21
6. **Combination** of 2 filters on **different columns**: Filter  for all white females that are 21 and younger and have less than 1000 `days_to_death`
http://localhost:8080/api/tdp/db/publicdb/tissue?filter_gender=female&filter_race=white&filter_lte_age=21&filter_lt_days_to_death=1000
7. **Combination with count**: The number of female from example 1.
http://localhost:8080/api/tdp/db/publicdb/tissue/count?filter_gender=female&filter_race=white&filter_lte_age=25

### Additionally 
* Multiple values for the suggested filter operator should be forbidden and throw an API error
* Throw API error for categorical values

example to exclude rows from 20-40:
```
filter range 1 = 10 ... 20
filter range 2 = 40 ... 50
```

* the current implementation only supports a single range
* multiple ranges are not supported -> this would require an `OR` condition
-> We dediced to not implement the ranges.

### Errors
A **Bad Request** respose with additional information will be shown for the following cases:
- The **column does not exist in the view**
**Request:** http://localhost:8080/api/tdp/db/publicdb/tissue?filter_lte_aeg=21
**Info Text:** Invalid filter key detected, "lte_aeg"
 --> because there is no 'aeg' column

- The **column type** for the the different greater and less filters **is not a number**
**Request:** http://localhost:8080/api/tdp/db/publicdb/tissue?filter_lte_gender=21
**Info Text:** Filters "lt","lte","gt", and "gte" are only applicable to columns of type "number", "gender" is not of type "number".
 --> because gender is a categorical column

- The **same filter is applied more than once** to on one column
**Request:** http://localhost:8080/api/tdp/db/publicdb/tissue?filter_lte_age=21&filter_lte_age=22
**Info Text:** Filter "lte_age" has too many values ("21", "22"), only one is allowed.
 --> because the less than equal filter is applied twice on the age column

- A **filter and its compement filter** is applied to the same column. Here a complement filter means the other filter from the same pair.
**Less pair:** `filter_lt` and `filter_lte`
**Greater pair:** `filter_gt` and `filter_gte`
**Request:** http://localhost:8080/api/tdp/db/publicdb/tissue?filter_lt_age=21&filter_lte_age=22 
**Info Text:** Filter "lte_age" has a complement filter "lt_age", only on of these filters is allowed.
 --> because both filters of the 'Less pair' are applied to the same column

### PR DESCRIPTION
**Pull request by PatrickAdelberger**
Fixes #187

This PR adds the less and greater filters (`filter_lt`, `filter_lte`, `filter_gt`, and `filter_gte`) to the API

### Possible Breaking Changes
The `filter_logic` function never raised an exception and only removed columns not included in the database view from the filter query.
- **Change:** The check for a key (column) now raises an exception if it is not included in the database view
```py
if not view.is_valid_filter(key):
  raise RuntimeError('Invalid filter key detected, "' + original_key + '"')
```
- **Change:** Additional exceptions can be raised because of the parameter/configuration checks for the new filters
```py
# check if column type is number for one of the greater (gt and gte) or less (lt and lte) filters
column_type = view.columns.get(key).get('type')
if is_greater_less_filter and column_type != 'number':
  raise RuntimeError('Filters "lt","lte","gt", and "gte" are only applicable to columns of type "number", "' + key + '" is not of type "number".')

# check if a greater (gt or gte) or less (lt or lte) filter was used on the same column more than once
if is_greater_less_filter and (len(where_clause[original_key]) > 1):
  separator = '", "'
  raise RuntimeError('Filter "' + original_key + '" has too many values ("' + separator.join(where_clause[original_key]) + '"), only one is allowed.')

# check complement filter for gte or lte (gt or lt respectively)
if check_complement_filter:
  complement_filter = original_key[:2]+'_'+key  # create complement filter
  complement_filter_exist = complement_filter in where_clause  # look if complement filter exist in the where clause
  # check if complement filter exists
  if complement_filter_exist:
    raise RuntimeError('Filter "' + original_key + '" has a complement filter "' + complement_filter + '", only on of these filters is allowed.')
```
#### Additionally
- **Change:** All `filter_logic` function calls are now placed within a try/except block, and create a 'Bad Request' response for the wrong filter configuration/parameters
```py
try:
  replacements, processed_args, extra_args, where_clause = filter_logic(view, args)
except RuntimeError as error:
  abort(400, error.message)
```

### CODE REVIEW

**Comment by lehnerchristian** [comment]
hi patrick! is there an app + plugin which I can check out and test this PR with?

**Comment by PatrickAdelberger** [comment]
hi christian!
yes, you can check the functionality in the develop branch of the cohort_product (https://github.com/Caleydo/cohort_product/tree/develop)

**Comment by thinkh** [comment]
@PatrickAdelberger Could you please fix the flake8 errors (see CircleCI log):

```
./tdp_core/sql_filter.py:118:18: E261 at least two spaces before inline comment
./tdp_core/sql_filter.py:122:18: E261 at least two spaces before inline comment
./tdp_core/sql_filter.py:124:1: W293 blank line contains whitespace
./tdp_core/sql_filter.py:129:18: E261 at least two spaces before inline comment
./tdp_core/sql_filter.py:131:32: E261 at least two spaces before inline comment
./tdp_core/sql_filter.py:134:18: E261 at least two spaces before inline comment
./tdp_core/sql_filter.py:136:32: E261 at least two spaces before inline comment
./tdp_core/sql_filter.py:146:20: E261 at least two spaces before inline comment
./tdp_core/sql_filter.py:146:21: E262 inline comment should start with '# '
```

**Comment by PatrickAdelberger** [comment]
@thinkh Sorry, I forgot about the failed build, but I fixed it now

**Comment by PatrickAdelberger** [comment]
change the fitler names:
less = filter_lt
greater = filter_gt
less equaly = filter_lte
greater equals = filter_gte

**Comment by PatrickAdelberger** [comment]
include error for less/greater filter on category column

**Comment by thinkh** [comment]
State of the in-person discussion:

* Multiple values for the suggested filter operator should be forbidden and throw an API error
* Throw API error for categorical values

example to exclude rows from 20-40:
```
filter range 1 = 10 ... 20
filter range 2 = 40 ... 50
```

* the current implementation only supports a single range
* multiple ranges are not supported -> this would require an `OR` condition

**Comment by PatrickAdelberger** [comment]
@lehnerchristian Could you look through the changes an merge this pull request if it is OK?
I updated this PR according to the specification we established during the code review.

**Comment by lehnerchristian** [comment]
hi, @PatrickAdelberger ! sorry, that it took me so long...

I've tested your PR a bit with Ordino. There was a change necessary for the query I've tested (see below). In my test I've discovered a little problem where you get the columns from the DBView and then get the type, because if no columns are defined in the DBView you'll get `null` and then the chained get with throw an error --> I'll create a suggestion

**change I've made to test the following URLs**:
add the following two lines below the following line https://github.com/Caleydo/tdp_publicdb/blob/develop/tdp_publicdb/query_score.py#L67
```
b.column('relativecopynumber', label='relativecopynumber', type='number')
b.filter('relativecopynumber')
```

**Test URLs**

After this fix I've tested it with the following queries:
- http://localhost:8080/api/tdp/db/publicdb/gene_cellline_score/score?table=copynumber&data_subtype=relativecopynumber&agg=avg&species=human&target=Ensembl&filter_panel_ensg=Cancer+Gene+Census&filter_gt_relativecopynumber=2
- http://localhost:8080/api/tdp/db/publicdb/gene_cellline_score/score?table=copynumber&data_subtype=relativecopynumber&agg=avg&species=human&target=Ensembl&filter_panel_ensg=Cancer+Gene+Census&filter_lt_relativecopynumber=2
- http://localhost:8080/api/tdp/db/publicdb/gene_cellline_score/score?table=copynumber&data_subtype=relativecopynumber&agg=avg&species=human&target=Ensembl&filter_panel_ensg=Cancer+Gene+Census&filter_gte_relativecopynumber=1.99972276035201
- http://localhost:8080/api/tdp/db/publicdb/gene_cellline_score/score?table=copynumber&data_subtype=relativecopynumber&agg=avg&species=human&target=Ensembl&filter_panel_ensg=Cancer+Gene+Census&filter_lte_relativecopynumber=1.99972276035201

** Result**
the filters seem to work fine. I get the same results for the GT and GTE as well as LT and LTE filters, but I've not gone into details --> the query has the correct operators. I only think the error message in https://github.com/datavisyn/tdp_core/blob/padelber/LessGreaterFilter/tdp_core/sql_filter.py#L161 needs to be improved, because it is also possible that the DBView has `.derive_columns()` that the types are derived automatically which does not work if 2 tables are joined together, hence the addition of `b.column('relativecopynumber', label='relativecopynumber', type='number')` in my changes above. So we should add the hint that if tables are joined together the column type needs to be added explicitly

File: tdp_core/sql_filter.py [hunk 1]
```diff
@@ -108,19 +108,70 @@ def to_clause(k, v):
...
+      check_complement_filter = True
+
+    # check if key (attribute/column) does exist in view
     if not view.is_valid_filter(key):
-      _log.warn('invalid filter key detected for view "%s" and key "%s"', view.query, key)
-      del where_clause[key]
+      raise RuntimeError('Invalid filter key detected, "' + original_key + '"')
+
+    # check if column type is number for one of the greater (gt and gte) or less (lt and lte) filters
+    column_type = view.columns.get(key).get('type')
```
Comment (lehnerchristian) [inline tdp_core/sql_filter.py]:
```suggestion
    column_type = view.columns.get(key, {}).get('type')
```

**Comment by lehnerchristian** [comment]
@thinkh , please also have a look if you find anything

**Pull request merged**